python/cuml/preprocessing/encoders.py

```diff
@@ -0,0 +1,291 @@
+        into a neural network or an unregularized regression.
+        - None : retain all features (the default).
+        - 'first' : drop the first category in each feature. If only one
+          category is present, the feature will be dropped entirely.
+        - Dict : ``drop[col]`` is the category in feature col that
+          should be dropped.
+    sparse : bool, default=True
+        Transform will return sparse matrix if True else will return an array.
+    dtype : number type, default=np.float
+        Desired datatype of transform's output.
+    handle_unknown : {'error', 'ignore'}, default='error'
+        Whether to raise an error or ignore if an unknown categorical feature
+        is present during transform (default is to raise). When this parameter
+        is set to 'ignore' and an unknown category is encountered during
+        transform, the resulting one-hot encoded columns for this feature
+        will be all zeros. In the inverse transform, an unknown category
+        will be denoted as None.
+
+    Attributes
+    ----------
+    drop_idx_ : array of shape (n_features,)
+        ``drop_idx_[i]`` is the index in ``categories_[i]`` of the category to
+        be dropped for each feature. None if all the transformed features will
+        be retained.
+    """
+    def __init__(self, categories='auto', drop=None, sparse=True,
+                 dtype=np.float, handle_unknown='error'):
+        self.categories = categories
+        self.sparse = sparse
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+        self.drop = drop
+        self._fitted = False
+        self.drop_idx_ = None
+        self._encoders = None
+        if sparse and np.dtype(dtype) not in ['f', 'd', 'F', 'D']:
+            raise ValueError('Only float32, float64, complex64 and complex128 '
+                             'are supported when using sparse')
+
+    def _validate_keywords(self):
+        if self.handle_unknown not in ('error', 'ignore'):
+            msg = ("handle_unknown should be either 'error' or 'ignore', "
+                   "got {0}.".format(self.handle_unknown))
+            raise ValueError(msg)
+        # If we have both dropped columns and ignored unknown
+        # values, there will be ambiguous cells. This creates difficulties
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` must be 'error' when the drop parameter is "
+                "specified, as both would create categories that are all "
+                "zero.")
+
+    def _check_is_fitted(self):
+        if not self._fitted:
+            raise RuntimeError("Model must first be .fit()")
+
+    def _compute_drop_idx(self):
+        if self.drop is None:
+            return None
+        elif isinstance(self.drop, str) and self.drop == 'first':
+            return {feature: cp.array([0])
```

Comment:
I don't think we need the index to be on the GPU

Reply:
As mentioned in the next comment below (for L139), I don't think there is a benefit going from GPU to CPU in the below case.

Here it is different indeed, as I'm creating a cupy array from the CPU. Although, while `return {feature: 0` works, I believe it is dangerous to have a function return an object either on GPU or on CPU depending on the branching, especially for further maintainability. Therefore, because of the below case that returns an object on GPU, I think it's better to keep it to `cp.array([0])` here.

Reply:
I would still let it be `return {feature: 0}`. Case in point with further maintainability, it adds unnecessary complexity where a person who has not written this code would have to follow and figure out why it's `cp.array([0])`, while the former is very readable

Reply:
Fair point, let's update to `return {feature: 0}`.